% -----------------------------------------------------------------------------
\section{Recommended Best Practices}
\label{sec:bestpractices}
% -----------------------------------------------------------------------------
\twozeroone{\subsection{SBOL Namespaces}

Namespaces for different versions of SBOL SHOULD NOT be semantically mixed in the same document. For example, an SBOL 2.x \sbol{ComponentInstance} SHOULD NOT refer to an SBOL 1.x DNAComponent. However, namespaces for different versions MAY be present in the same document so long as they are semantically independent (that is, so long as objects belonging to these namespaces do not refer to each other). When multiple SBOL namespaces are present in a document, libraries SHOULD default to interpreting as SBOL only those objects belonging to the namespace for the most recent version of SBOL in the document. Any remaining objects belonging to any other SBOL namespace SHOULD be interpreted as \sbol{GenericTopLevel}s and custom \sbol{Annotation}s.}

\subsection{Use of the Version Property}

Once an SBOL object has been published where others might have access it (e.g., to an online repository), it might be the case that other objects come to depend on the particular contents of the published object. 
Thus, in order to avoid creating conflicting data, if a person wants to change the properties of a published object, they SHOULD do so by making a new copy of object that incorporates the change and has an \sbol{identity} property that contains a new \sbol{URI}.

The relationship between the old and new objects (i.e., that the new object was derived from the old object), however, is not visible unless it is explicitly declared.  This is RECOMMENDED to be done using the \sbol{persistentIdentity}, and \sbol{version} properties. The preferred practice for declaring such a relationship is to use the same \sbol{persistentIdentity} for both objects, but give the newer object a later \sbol{version}. Then, when the new object is published, it can be clear to both humans and machines that this object is intended to update the previously published object. In this way, when a user wants the latest version of an object, they can obtain it by referencing the object via its \sbol{persistentIdentity} and rely on a tool to find the object with that \sbol{persistentIdentity} and the latest \sbol{version}.

As stated in \ref{sec:version},  it is RECOMMENDED that version numbering follow
the conventions of semantic versions (\url{http://semver.org/}), particularly as implemented by Maven (\url{http://maven.apache.org/}).
This convention represents versions as sequences of numbers and qualifiers separated by the characters {\tt .} and {\tt -} and compared in lexicographical order (for example, 1 < 1.3.1 < 2.0-beta).  For a full explanation, see the linked resources.

\subsection{Compliant SBOL Objects}
\label{sec:compliant}

Maintaining unique \sbol{identity} \sbol{URI}s for all SBOL objects can be a very challenging implementation task.  To reduce this burden, users of SBOL 2.x are encouraged to follow a few simple rules when constructing the \sbol{identity} properties and related properties for SBOL objects.  When these rules are followed in constructing an SBOL object, we say that this object is \emph{compliant}. These rules are as follows:
\begin{enumerate}
\item The \sbol{identity} of a compliant SBOL object MUST begin with a \emph{URI prefix} that maps to a domain over which the user has control. Namely, the user can guarantee uniqueness of identities within this domain.
\item The \sbol{persistentIdentity} and \sbol{displayId} properties are REQUIRED of a compliant SBOL object.
\item The \sbol{persistentIdentity} of a compliant \sbol{TopLevel} object MUST end with a delimiter (`/', `\#', or `:') followed by the \sbol{displayId} of the object. 
\item The \sbol{persistentIdentity} of a compliant SBOL object that is not also a \sbol{TopLevel} object MUST begin with the \sbol{persistentIdentity} of its parent object and be immediately followed by a delimiter (`/', `\#', or `:') and the \sbol{displayId} of the compliant object.
\item If a compliant SBOL object is not given a \sbol{version}, then its \sbol{identity} and \sbol{persistentIdentity} properties MUST contain the same \sbol{URI}.
\item If a compliant SBOL object has a \sbol{version}, then its \sbol{identity} property MUST contain a \sbol{URI} of the form  ``\refObj{persistentIdentity}/\refObj{version}''.
\item The \sbol{version} of a compliant SBOL object that is not also a \sbol{TopLevel} object MUST contain the same \sbol{String} as the \sbol{version} property of the compliant object's parent object.
\item The \sbol{identity}, \sbol{persistentIdentity}, \sbol{displayId}, and \sbol{version} properties of a compliant SBOL object MUST NOT be changed once set.
\end{enumerate}

All examples in this specification use compliant \sbol{URI}s.

\subsection{Annotations: Embedded Objects vs. External References}

When annotating an SBOL document with additional information, there are
two general methods that can be used:
\begin{itemize}
\item Embed the information in the SBOL document, either as non-SBOL
  properties or \sbol{GenericTopLevel} objects.
\item Store the information separately and annotate the SBOL document
  with \sbol{URI}s that point to it.
\end{itemize}
In theory, either method can be used in any case. (Note that a third case not
discussed here is to use SBOL to annotate external objects with linking
to SBOL documents, rather than annotate SBOL documents with links external objects.)

In practice, 
embedding massive amounts of non-SBOL data into SBOL documents is likely
to cause problems for people and software tools trying to manage and
exchange such documents.  Therefore, it is RECOMMENDED that small amounts of information (e.g., design notes or preferred graphical layout) be embedded in the SBOL model, while large amounts of information (e.g., the contents of the scientific publication from which a model was derived or flow cytometry data that characterizes performance) be linked with URIs pointing to external resources.  The boundary between ``small'' and ``large'' is left deliberately vague, recognizing that it will likely depend on the particulars of a given SBOL application.

\subsection{Completeness and Validation}

RDF documents containing serialized SBOL objects might or might not be
entirely self-contained.  A SBOL document is self-contained or ``complete'' if every SBOL object referred to in the document is contained in the document.  It is RECOMMENDED that serializations be complete whenever practical.  In order words, when serializing an SBOL object, serialize all of the other objects that it points to, then serialize all of the other objects that these objects point to, etc., until the document is complete.

It is important to note that there is no guarantee that an RDF document
contains valid SBOL. When an RDF document is deserialized to SBOL
objects, the program doing so SHOULD verify that all of the property
values encoded therein have the right data type (e.g., that the object
pointed to by the \sbol{sequences} property of a
\sbol{ComponentDefinition} is really a \sbol{Sequence}).
For complete files, this validation can be carried out entirely locally. For files that are not complete, an implementation either needs to
have a means of validating those external references (e.g., by
retrieving them from various repositories), or it needs to mark them as
unverified and not depend on their correctness.

\subsection{Recommended Ontologies for External Terms}
\label{sec:recomm_ontologies}

External ontologies and controlled vocabularies are an integral part of SBOL. SBOL uses \sbol{URI}s to access existing biological information through these resources. New SBOL specific terms are defined only when necessary. For example, \sbol{ComponentDefinition} \sbolmult{types:CD}{types}, such as DNA or protein, are described using BioPAX terms. Similarly, the roles of a DNA \twozeroone{or RNA} \sbol{ComponentDefinition} are described via SO terms. Although RECOMMENDED ontologies have been indicated in relevant sections where possible, other resources providing similar terms can also be used. A summary of these external sources can be found in \ref{tbl:preferred_external_resources}.

\begin{table}[ht]
  \begin{edtable}{tabular}{p{3cm}p{1.5cm}p{4.5cm}p{6cm}}
    \toprule
    \textbf{SBOL Entity} & \textbf{Property} & \textbf{Preferred External Resource}
    & \textbf{More Information} \\
    \midrule
    \textbf{ComponentDefinition}  & types & BioPAX & \url{http://www.biopax.org}\\
                                  & types & SO (nucleic acid topology)& \url{http://www.sequenceontology.org}\\
    						   	  & roles & SO (\textit{DNA} or \textit{RNA}) & \url{http://www.sequenceontology.org}   \\
    						   	  & roles & CHEBI (\textit{small molecule}) & \url{https://www.ebi.ac.uk/chebi/}   \\
%    						   	  & roles & UniProt (if type is \textit{protein}??) \\   
    \textbf{Interaction}	      & types & SBO (occurring entity branch) & 
    \url{http://www.ebi.ac.uk/sbo/main/} \\
    \textbf{Participation}	      & roles & SBO (participant roles branch) &
    \url{http://www.ebi.ac.uk/sbo/main/} \\
    \textbf{Model}	      		  & language & EDAM & \url{http://bioportal.bioontology.org/ontologies/EDAM}     \\
    				      		  & framework & SBO (modeling framework branch) &
    \url{http://www.ebi.ac.uk/sbo/main/} \\
    \bottomrule
  \end{edtable}
  \caption{Preferred external resources from which to draw values for various SBOL properties.}
  \label{tbl:preferred_external_resources}
\end{table}

\twoonezero{The URIs for ontological terms SHOULD come from identifiers.org.  However, it is acceptable to use terms from purl.org as an alternative, for example when RDF tooling requires URIs to be represented as compliant QNames.  SBOL software may convert between these forms as required.}

\twozeroone{\subsection{Annotating Entities with Date \& Time}\label{sec:DateTime}
Entities in an SBOL document can be annotated with creation and modification date. It is RECOMMENDED that predicates, or properties, from DCMI Metadata Terms SHOULD be used to include date and time information. The \texttt{created} and \texttt{modified} terms SHOULD respectively be used to annotate SBOL entities with creation and modification dates. Date and time values SHOULD be expressed using the XML Schema \texttt{DateTime} datatype~\citep{Biron2004}. For example, "\texttt{2016-03-16T20:12:00Z}" specifies that the day is 16 March 2016 and the time is 20:12pm in UTC (Coordinated Universal Time). An example serialization is shown below:}

\lstsetsbol
\begin{lstlisting}
<?xml version="1.0" ?>
<rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#" xmlns:dcterms="http://purl.org/dc/terms/" xmlns:prov="http://www.w3.org/ns/prov#" xmlns:sbol="http://sbols.org/v2#">
  <sbol:CLASS_NAME rdf:about="...">
    <dcterms:created>DATE_TIME_STRING</dcterms:created>
    <dcterms:modified>DATE_TIME_STRING</dcterms:modified>
               ...
  </sbol:CLASS_NAME>
  ...
</rdf:RDF>
\end{lstlisting}

\twotwoone{\subsection{Annotating Entities with Authorship information}\label{sec:Authorship}
Authorship information should ideally be added to \sbol{TopLevel} entities, where possible. It is RECOMMENDED that the \texttt{creator} DCMI Metadata term SHOULD be used to annotate SBOL entities with authorship information. This property can be repeated for each author. The example below shows the use of this property for two authors and the values shown are free text \texttt{String} literals.
}
\lstsetsbol
\begin{lstlisting}
<?xml version="1.0" ?>
<rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#" xmlns:dcterms="http://purl.org/dc/terms/" xmlns:prov="http://www.w3.org/ns/prov#" xmlns:sbol="http://sbols.org/v2#">
  <sbol:CLASS_NAME rdf:about="...">
    <dcterms:creator>First author</dcterms:creator>
    <dcterms:creator>Second author</dcterms:creator>
               ...
  </sbol:CLASS_NAME>
  ...
</rdf:RDF>
\end{lstlisting}

\twothreeone{\subsection{Host Context / Ontologies for Experiments}

\subsubsection{Mixtures via ModuleDefinitions}

Any \sbol{ModuleDefinition} can be interpreted as specifying a mixture of the \sbol{FunctionalComponent}s and \sbol{Module}s that it includes.

The amount of each such instance included in the mixture SHOULD be specified by attaching a \sbol{Measure} with a \sbolmult{types:Measure}{type} set to the appropriate SBO term. The SBO terms that are RECOMMENDED as appropriate are members of the Systems Description Parameter (SBO:0000545) branch of SBO. Examples include:
\begin{itemize}
\item SBO:0000540: fraction of an entity pool (e.g., 1/3 CHO cells, 2/3 HEK cells)
\item SBO:0000472: molar concentration of an entity (e.g., 1 mM Arabinose)
\item SBO:0000361: amount of an entity pool (e.g., 200 uL M9 media)
\end{itemize}

\subsubsection{Media, Inducers, and Other Reagents}

An ``atomic'' reagent, which is not expected to be divided further in designs, SHOULD be represented as a \linebreak \sbol{ComponentDefinition}. Examples: Arabinose, Kanamycin, Rainbow bead control

A ``composite'' reagent, which might be defined as a mixture of other reagents, SHOULD be represented as a \sbol{ModuleDefinition} that defines it as a mixture of other reagents. This might be recursive, with mixtures of mixtures of mixtures, etc. Examples: M9 media, LB broth

The roles of reagents may vary in context: for example, Arabinose may serve as an inducer or as a media carbon source. As such, role SHOULD be indicated by an NCI Thesaurus (NCIT) term in a \sbolmult{roles:MD}{role} property of the \sbol{FunctionalComponent} or \sbol{Module}. Examples include:
\begin{itemize}
\item NCIT:C64356: Positive Control
\item NCIT:C48694: Cell
\item NCIT:C85504: Media
\item NCIT:C14419: Strain
\item NCIT:C120268: Inducer
\end{itemize}

Note that this also implies that the \sbolmult{roles:C}{role} property should be promoted from \sbol{Component} to \sbol{ComponentInstance} (so it can be used for \sbol{FunctionalComponent} as well) and also added to \sbol{Module}.

\subsubsection{Strains, Genomes, and Plasmids}

A cell strain SHOULD be represented as a \sbol{ModuleDefinition} containing at least:
\begin{itemize}
\item A \sbol{FunctionalComponent} for the cell, which is RECOMMENDED to have \sbolmult{types:CD}{type} cell (GO:0005623) and \sbolmult{roles:MD}{role} physical compartment (SBO:0000290).
\item A representation of its genetic material (genome, plasmids, etc.) as a collection of \sbol{FunctionalComponent} and/or \sbol{Module} instances.
\item An \sbol{Interaction} of type \texttt{FunctionalCompartment} (SBO:0000289) with the cell participating as physical compartment (SBO: 0000290) and the genetic material participating as contained entity (SBO:0000664).
\end{itemize}

\subsubsection{Samples}

A complete specification of a sample SHOULD be a \sbol{ModuleDefinition} that includes at least:
\begin{itemize}
\item A \sbol{Module} instantiating each strain in the sample
\item A \sbol{Module} or \sbol{FunctionalComponent} for the media or buffer
\item A \sbol{Module} or \sbol{FunctionalComponent} for each additional reagent added to the media (e.g., inducers, antibiotics)
\item \sbol{Measure}s on each of these specifying the amount in the sample
\item \sbol{Measure}s on the \sbol{ModuleDefinition} for each environmental parameter (e.g., temperature, pH, culturing time)
\end{itemize}

\subsubsection{Other Experimental Parameters}

In order to deal with parameters associated with the context in general but not specific instances, e.g., temperature, pH, total sample volume, we need to add a \texttt{measures} property to \sbol{ModuleDefinition} (in addition to the ones already on \sbol{Module} and \sbol{ComponentInstance}) and to \sbol{ComponentDefinition}.

The distinction is that the \texttt{measures} of a \texttt{Definition} provide context-free information (e.g., the pH of M9 media, the GC-content of a GFP coding sequence), while the \texttt{measures} on an \texttt{Instance} provide a measurement in context (e.g., the dosage of Arabinose in a sample).

Values of these parameters SHOULD be specified by attaching a \sbol{Measure} with a \sbolmult{types:Measure}{type} set to the appropriate SBO term. The SBO terms that are RECOMMENDED as appropriate are members of the Systems Description Parameter (SBO:0000545) branch of SBO. Examples include:
\begin{itemize}
\item SBO:0000147: thermodynamic temperature (e.g., culturing at 27 C)
\item SBO:0000332: half-life of an exponential decay (e.g., decay rate of a gRNA)
\item SBO:0000304: pH (e.g., pH of M9 media)
\end{itemize}}

\twothreeone{\subsection{Multicellular System Designs}

SBOL has been used extensively to represent designs in homogeneous systems, where the same design is implemented in every cell. However, in recent years there has been increasing interest in multicellular systems, where biological designs are split across multiple cells to optimize the system behavior and function. Therefore, there is a need to define a set of best practices so that multicellular systems can be captured using SBOL in a standard way.

\subsubsection{Representing Cells}

To represent multicellular systems using SBOL, it is first necessary to represent cells properly. When doing so, it is important to be able to capture the following information: (i) taxonomy, (ii) interactions occurring within the cell, and (iii) components inside the cell (e.g. DNA Molecules). The proposed approach described in this section, depicted in \ref{uml:cell_representation}, is capable of capturing this information. It uses an instance of the \sbol{ModuleDefinition} class to represent a system which contains the cell. The cell is represented by an instance of the \sbol{FunctionalComponent} class inside the \sbol{ModuleDefinition}. A separate \sbol{ComponentDefinition} instance is used to capture information about the species and strain of the cell in the design. This \sbol{ComponentDefinition} has a \sbolmult{types:CD}{type} of ’cell’ from the Gene Ontology (GO:0005623), and a \sbolmult{roles:CD}{role} of ’physical compartment’ (SBO:0000290). Taxonomic information is captured by annotating the class instance with a URI for an entry in the NCBI Taxonomy Database. The \sbol{ComponentDefinition} instance is used as the value of the \sbolmult{definition:CI}{definition} property of the \sbol{FunctionalComponent} which represents the cell in the design to specify information about its taxonomy. Other entities which are relevant to this aspect of the design are also captured using instances of the \sbol{FunctionalComponent} class. Interactions which occur in this system are captured using the \sbol{Interaction} and \sbol{Participation} classes. Interactions which occur within the cell are specified by \sbol{Interaction} classes which contain the \sbol{FunctionalComponent} instance representing the cell as a \sbol{participant} with a \sbolmult{roles:P}{role} of 'physical compartment'.

With this approach, an additional \sbol{Interaction} class instance could be used to define entities which are only present within the cell, i.e. they are not available to the rest of the system. This interaction has a \sbolmult{types:I}{type} of ’containment’ (SBO:0000469), and has at least two participants. One of these \sbol{participant}s is the cell, which has a \sbolmult{roles:P}{role} of ’physical compartment’, whilst the others are the contained entities, which have \sbolmult{roles:P}{roles} of ’contained’
(SBO:0000064).

\begin{figure}[ht]
	\begin{center}
		\includegraphics[width=\textwidth]{uml/cell_representation}
		\caption[Repressenting a cell]{This is a proposed approach for capturing cell designs in SBOL. A \sbol{ComponentDefinition} annotated with a URI pointing to an entry in the NCBI Taxonomy Database is used to capture information about the cell's strain/species. The \sbol{ComponentDefinition} instance has a type of 'Cell' from the Gene Ontology (GO), and a role of 'physical compartment'. An instance of the \sbol{ModuleDefinition} class is used to represent a system in which the cell is implemented. Entities, including the cell, are instantiated as \sbol{FunctionalComponent}s, and process are captured using the \sbol{Interaction} class. Process which are contained within the cell are represented by including the cell as a participant with a role of 'physical compartment'. C) Example of how the \sbol{Interaction} class can be used to explicitly confer that an entity is contained inside a cell. Here a plasmid is specified as being contained within a cell.
		}
		\label{uml:cell_representation}
	\end{center}
\end{figure}

\subsubsection{Multiple Cells in a Single Design}

The approach to represent multiple cells in SBOL uses both \sbol{Module} and \sbol{FunctionalComponent} classes to represent cells, and any non-cell entities must also be instantiated as \sbol{FunctionalComponent} classes. 
The \sbol{Module} instances are defined by the \sbol{ModuleDefinition} classes used to represent cells/cell systems as in the first approach. \sbol{FunctionalComponent} classes are also used to represent taxonomic information about the cell, and are defined by the \sbol{ComponentDefinition} classes which capture this information. Finally, the \sbol{Module} classes contain instances of the \sbol{MapsTo} class, which are used to explicitly capture links between the same entities present in multiple parts of the same design. Here, a \sbol{MapsTo} class with a \sbol{refinement} value of ’merge’ is used to link \sbol{FunctionalComponent} classes which represent a cell in the multicellular system to the \sbol{FunctionalComponent} class used to represent the same cell in the lower-level cell system design. Instances of the \sbol{MapsTo} class are also used to capture that non-cell entities in the multicellular system are identical to entities used in the cell system design. This is demonstrated in \ref{uml:multiple_cell_representation}.

\begin{figure}[ht]
	\begin{center}
		\includegraphics[width=\textwidth]{uml/Approach3}
		\caption[]{Captured here is a design involving two cells which both interact with the small molecule 'Molecule A'. Designs for Cell 1 and Cell are captured using the approach depicted in \ref{uml:cell_representation}. The overall multicellular system is represented by a \sbol{ModuleDefinition} with a \sbolmult{roles:MD}{role} of 'functional compartment', which is an SBO term. The two systems involving Cell 1 and Cell 2 are included in this multicellular design as instances of the \sbol{Module} class. Entities in the multicellular design, including those in the Cell 1 and Cell 2 system designs, are instantiated as a \sbol{FunctionalComponent}. Instances of the \sbol{MapsTo} class are used to map entities specified in both the individual cell systems, and the multicellular system.
		}
		\label{uml:multiple_cell_representation}
	\end{center}
\end{figure}

\subsubsection{Cell Ratios}

The proportion of cell types present in a multicellular system has been captured through the use of an instance of the \sbol{Measure} class on classes which represent cell systems in the multicellular design. As a best practice, the value of these measure classes is a number less than or equal to 100\%. These values are the percentage of a cell type present in the system compared to all other cell types present. Therefore, the sum of all these values specified in the system should be equal to 100\%. These approaches are shown in \ref{uml:cell_ratios}.

\begin{figure}[ht]
	\begin{center}
		\includegraphics[width=\textwidth]{uml/Cell_Ratios}
		\caption[]{Annotating class instances with cellular proportions. Instances of the Measure class are used to capture the percentage of each cell type present in the multicellular system design.
		}
		\label{uml:cell_ratios}
	\end{center}
\end{figure}
}